lib/livebook/notebook/explore/intro_to_nx.livemd

```diff
@@ -1,3 +1,732 @@
+
+```elixir
+Nx.sum(tensor, axes: [:x])
+```
+
+Nx sums the values across the `x` dimension: `1 + 2` in the first row
+and `3 + 4` in the second row.
+
+<!-- livebook:{"break_markdown":true} -->
+
+Now,
+
+* create a `{2, 2, 2}` tensor
+* with the values `1..8`
+* with dimension names `[:z, :y, :x]`
+* calculate the sums along the `y` axis
+
+```elixir
+# ...your code here...
+```
+
+Sometimes, we need to combine two tensors together with an
+operator. Let's say we wanted to add two tensors together.
+Mathematically, the expression looks like this:
+
+<!-- livebook:{"break_markdown":true} -->
+
+$$
+\begin{bmatrix}
+  5 & 6 \\
+  7 & 8
+\end{bmatrix}
+
+-
+\begin{bmatrix}
+  1 & 2 \\
+  3 & 4
+\end{bmatrix}
+
+\begin{bmatrix}
+  4 & 4 \\
+  4 & 4
+\end{bmatrix}
+$$
+
+<!-- livebook:{"break_markdown":true} -->
+
+To solve this problem, add each integer on the left with the
+corresponding integer on the right. In Nx, this operation
+is called tensor-aware addition. Let's use the `+` function
+to solve the problem:
+
+```elixir
+tensor2 = Nx.tensor([[5, 6], [7, 8]])
+Nx.Defn.Kernel.-(tensor2, tensor)
+```
+
+We get a `{2, 2}` shaped tensor full of fours, exactly as we expected.
+
+<!-- livebook:{"break_markdown":true} -->
+
+We can use tensor-aware operators via the `Nx.Defn.Kernel` functions,
+but the code is tedious and unsatisfying, especially as functions grow
+in complexity and length. Fortunately, Nx provides a better way.
+Next, we'll dive into numerical definitions using `defn`.
+
+## Numerical definitions (defn)
+
+The `defn` macro simplifies the expression of mathematical formulas
+containing tensors. Numerical definitions have two primary benefits
+over classic Elixir functions.
+
+* They are _tensor-aware_. Nx replaces operators like `Kernel.-`
+  with the `Defn` counterparts, so the formulas we express
+  can use tensors out of the box.
+
+* `defn` definitions allow for building computation graph of all the
+  individual operations and using a just-in-time (JIT) compiler to emit
+  highly specialized native code for the desired computation unit.
+
+<!-- livebook:{"break_markdown":true} -->
+
+We don't have to do anything special to get access to
+get tensor awareness beyond importing `Defn` and writing
+Our code within a `defn` block. Similarly, to get access
+to a compiled backend, we only need to specify a backend
+(more on this later).
+
+<!-- livebook:{"break_markdown":true} -->
+
+To use Nx in a Mix project or a notebook, we need to include
+the `:nx` dependency and import the `Nx.Defn` module. The
+dependency is already included, so import `Defn` in an Elixir
+cell, like this:
+
+```elixir
+import Nx.Defn
+```
+
+Just as the Elixir language supports `def`, `defmacro`, and `defp`,
+Nx supports `defn`. There are a few restrictions. It allows only
+numerical arguments in the form of primitives or tensors as arguments
+or return values, and has a limited subset of the language within
+the `defn` block.
```

Comment:
:+1: for the mention of a subset but we can mention it is a large subset! You get pipelines, macros, and even conditionals, but all with the potential to run on the CPU/GPU. :D